text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,527 @@
+    println!("{}", bt);
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+In Rust, errors typically gather two forms of context when they are created:
+context for the *current error message* and context for the *final* *error
+report*. The `Error` trait exists to provide an interface to context intended
+for error reports. This context includes the error message, the source error,
+and, more recently, backtraces.
+
+However, the current approach of promoting each form of context to a method on
+the `Error` trait doesn't leave room for forms of context that are not commonly
+used, or forms of context that are defined outside of the standard library.
+
+## Extracting non-std types from `dyn Errors`
+
+By adding a generic form of these member access functions we are no longer
+restricted to types defined in the standard library. This opens the door to
+many new forms of error reporting.
+
+### Example use cases this enables
+
+* using alternatives to `std::backtrace::Backtrace` such as
+  `backtrace::Backtrace` or [`SpanTrace`]
+* zig-like Error Return Traces by extracting `Location` types from errors
+  gathered via `#[track_caller]` or similar.
+* error source trees instead of chains by accessing the source of an error as a
+  slice of errors rather than as a single error, such as a set of errors caused
+  when parsing a file TODO reword
+* Help text such as suggestions or warnings attached to an error report
+
+## Moving `Error` into `libcore`
+
+Adding a generic member access function to the `Error` trait and removing the
+`backtrace` function would make it possible to move the `Error` trait to libcore
+without losing support for backtraces on std. The only difference being that
+in places where you can currently write `error.backtrace()` on nightly you
+would instead need to write `error.context::<Backtrace>()`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Error handling in Rust consists of three steps: creation/propagation, handling,
+and reporting. The `std::error::Error` trait exists to bridge the gap between
+creation and reporting. It does so by acting as a interface that all error
+types can implement that defines how to access context intended for error
+reports, such as the error message, source, or location it was created. This
+allows error reporting types to handle errors in a consistent manner when
+constructing reports for end users while still retaining control over the
+format of the full report.
+
+The `Error` trait accomplishes this by providing a set of methods for accessing
+members of `dyn Error` trait objects. It requires that types implement the
+display trait, which acts as the interface to the main member, the error
+message itself.  It provides the `source` function for accessing `dyn Error`
+members, which typically represent the current error's cause. It provides the
+`backtrace` function, for accessing a `Backtrace` of the state of the stack
```

Comment:
Is it worth mentioning that backtrace is unstable?

Reply:
FWIW realizing that this could subsume `Error::backtrace` is waaaay cool.

Reply:
it could also subsume `source` fwiw :D

Reply:
True, but we can't *delete* `source`. I guess deprecation markers have been pretty successful in the ecosystem.

Reply:
yea, i don't really think its necessary to delete source, at a certain point it is okay to optimize your API for the happy path, and sources are by far the most common kind of context, and its only adding a dependency on its own definition, so its presence not an issue as far as moving the Error trait around is concerned afaik.